### PR TITLE
Added InternalServerError and InternalServerError<TValue> to TypedResults

### LINKED ISSUE #53073: Add InternalServerError and InternalServerError<T> to TypedResults
## Summary

As a .NET developer who develops a lot of Web API:s, I often use the new TypedResults with return type Result<T1, T2, ..>. In every project that I use this, I find myself implementing a custom return type of ```InternalServerError``` and ```InternalServerError<TValue>``` since it's missing in the standard implementation. 

## Motivation and goals

If I'm using TypedResults, I'm doing that for two major reasons:

- Compile time errors for return types that doesn't match the intended design (specified in the return type of the method)
- Autogeneration of metadata for OpenAPI specification

## In scope

- You want to be able to have a 500 Internal Server error containing a complex type as one of your possible return types for your endpoint and get compile time errors if it doesn't match what you're actually returning.
- You want the return type of 500 Internal Server Error containing the complex type to be described in your autogenerated OpenAPI specification, so that your clients knows what to expect. 
- Same as above, but a 500 Internal Server Error with no content.

## Risks / unknowns

This can be misused as a general return type for client errors, but I feel like that is more of a violation of REST rather than the framework.

## Examples

Simplified example:

```cs
app.MapGet("/books/{id}", async Task<Results<Ok<Book>, NotFound<string>, InternalServerError<string>>> (int id) => {
  
    try
    {
        // Retrieve the book
        var book = await ... 

        if (book is null)
            return TypedResults.NotFound($"Could not find book with id: {id}");

        // Found book
        return TypedResults.Ok(book);
    }
    catch (Exception ex)
    {
        Log.Error($"Something unexpected happened: {ex}");
        return TypedResults.InternalServerError("Something unexpected happened while processing your request, please contact the API provider");
    }

});
```

# Detailed design

Looking at the other TypedResults, I would implement this as follows:

```cs
/// <summary>
/// An <see cref="IResult"/> that on execution will write an object to the response
/// with Internal Server Error (500) status code.
/// </summary>
public sealed class InternalServerError : IResult, IEndpointMetadataProvider, IStatusCodeHttpResult
{
    /// <summary>
    /// Initializes a new instance of the <see cref="InternalServerError"/> class with the values
    /// provided.
    /// </summary>
    internal InternalServerError()
    {
    }

    /// <summary>
    /// Gets the HTTP status code: <see cref="StatusCodes.Status500InternalServerError"/>
    /// </summary>
    public int StatusCode => StatusCodes.Status500InternalServerError;

    int? IStatusCodeHttpResult.StatusCode => StatusCode;

    /// <inheritdoc/>
    public Task ExecuteAsync(HttpContext httpContext)
    {
        ArgumentNullException.ThrowIfNull(httpContext);

        // Creating the logger with a string to preserve the category after the refactoring.
        var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
        var logger = loggerFactory.CreateLogger("Microsoft.AspNetCore.Http.Result.InternalServerErrorObjectResult");

        HttpResultsHelper.Log.WritingResultAsStatusCode(logger, StatusCode);
        httpContext.Response.StatusCode = StatusCode;

        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    static void IEndpointMetadataProvider.PopulateMetadata(MethodInfo method, EndpointBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(method);
        ArgumentNullException.ThrowIfNull(builder);

        builder.Metadata.Add(new ProducesResponseTypeMetadata(StatusCodes.Status500InternalServerError, typeof(void)));
    }
}
```

.. and

```cs
/// <summary>
/// An <see cref="IResult"/> that on execution will write an object to the response
/// with Internal Server Error (500) status code.
/// </summary>
/// <typeparam name="TValue">The type of error object that will be JSON serialized to the response body.</typeparam>
public sealed class InternalServerError<TValue> : IResult, IEndpointMetadataProvider, IStatusCodeHttpResult, IValueHttpResult, IValueHttpResult<TValue>
{
    /// <summary>
    /// Initializes a new instance of the <see cref="InternalServerError"/> class with the values
    /// provided.
    /// </summary>
    /// <param name="error">The error content to format in the entity body.</param>
    internal InternalServerError(TValue? error)
    {
        Value = error;
        HttpResultsHelper.ApplyProblemDetailsDefaultsIfNeeded(Value, StatusCode);
    }

    /// <summary>
    /// Gets the object result.
    /// </summary>
    public TValue? Value { get; }

    object? IValueHttpResult.Value => Value;

    /// <summary>
    /// Gets the HTTP status code: <see cref="StatusCodes.Status500InternalServerError"/>
    /// </summary>
    public int StatusCode => StatusCodes.Status500InternalServerError;

    int? IStatusCodeHttpResult.StatusCode => StatusCode;

    /// <inheritdoc/>
    public Task ExecuteAsync(HttpContext httpContext)
    {
        ArgumentNullException.ThrowIfNull(httpContext);

        // Creating the logger with a string to preserve the category after the refactoring.
        var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
        var logger = loggerFactory.CreateLogger("Microsoft.AspNetCore.Http.Result.InternalServerErrorObjectResult");

        HttpResultsHelper.Log.WritingResultAsStatusCode(logger, StatusCode);
        httpContext.Response.StatusCode = StatusCode;

        return HttpResultsHelper.WriteResultAsJsonAsync(
                httpContext,
                logger: logger,
                Value);
    }

    /// <inheritdoc/>
    static void IEndpointMetadataProvider.PopulateMetadata(MethodInfo method, EndpointBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(method);
        ArgumentNullException.ThrowIfNull(builder);

        builder.Metadata.Add(new ProducesResponseTypeMetadata(StatusCodes.Status500InternalServerError, typeof(TValue), new[] { "application/json" }));
    }
}
```

I would open a PR, but the code of conduct said to post a design proposition first.

### PR DESCRIPTION
**Pull request by onurmicoogullari**
# Extending TypedResults with InternalServerError

- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description
Added ``InternalServerError`` and ``InternalServerError<TValue>`` to ``TypedResults``.

Fixes #53073

### CODE REVIEW

**Comment by onurmicoogullari** [comment]
> @onurmicoogullari please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

**Comment by davidfowl** [comment]
APIs need to go through APi review and have to be approved by the team.

File: src/Http/Http.Results/src/Results.cs [hunk 1]
```diff
@@ -690,6 +691,29 @@ public static IResult UnprocessableEntity(object? error = null)
...
+    /// <returns>The created <see cref="IResult"/> for the response.</returns>
+    public static IResult InternalServerError<TValue>(TValue? error)
+        => error is null ? TypedResults.InternalServerError() : TypedResults.InternalServerError(error);
+
+    /// <summary>
+    /// Produces a <see cref="StatusCodes.Status500InternalServerError"/> response.
+    /// </summary>
+    /// <param name="exception">The exception to be thrown.</param>
+    /// <returns>The created <see cref="IResult"/> for the response.</returns>
+    public static IResult InternalServerError(ExceptionDispatchInfo exception)
```
Comment (halter73) [inline src/Http/Http.Results/src/Results.cs:714]: We just [approved the API](https://github.com/dotnet/aspnetcore/issues/53073#issuecomment-1961936888) with everything except these `ExceptionDispatchInfo` overloads. I know it was our suggestion in the first place, so sorry to give you the runaround.

Thanks for sticking with this as we went through the API review process.

Comment (halter73) [reply src/Http/Http.Results/src/Results.cs:714]: Yes. Thank you!

Comment (onurmicoogullari) [reply src/Http/Http.Results/src/Results.cs]: Done!

**Review by halter73 — APPROVED** [review]

**Pull request merged**